### PR TITLE
Show poll question when rendering reply to a poll end event (PSG-1133)

### PR DESCRIPTION
**Pull request by onurays**
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

We recently started to render a message when a poll is ended and users are able to reply ended polls. We need to render poll question for ended poll messages.

## Motivation and context

Instead of rendering "In reply to ended poll" we will render "In reply to {poll_question}"

## Tests

- Create a poll
- End a poll
- Reply to the ended poll message
- See the replied message is rendered by including the poll question

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt [hunk 1]
```diff
@@ -612,7 +612,7 @@ internal class LocalEchoEventFactory @Inject constructor(
         val userId = eventReplied.root.senderId ?: return null
         val userLink = permalinkFactory.createPermalink(userId, false) ?: return null
 
-        val body = bodyForReply(eventReplied.getLastMessageContent(), eventReplied.isReply(), isRedactedEvent)
+        val body = bodyForReply(eventReplied, isRedactedEvent)
```
Comment (mnaturel) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt:615]: I see there is the same logic in `createReplaceTextOfReply` method for the case of editing a reply message. Maybe we should do the same modification as here, no?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt [hunk 2]
```diff
@@ -724,6 +724,18 @@ internal class LocalEchoEventFactory @Inject constructor(
         }
     }
 
+    private fun bodyForReply(timelineEvent: TimelineEvent, isRedactedEvent: Boolean = false): TextContent {
+        val content = when (timelineEvent.root.getClearType()) {
+            in EventType.POLL_END.values -> {
+                localEchoRepository
```
Comment (mnaturel) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt:730]: I think we should have a fallback for when the related poll event is not found in the repository.
```suggestion
                localEchoRepository
                        .getRelatedPollEvent(timelineEvent)
                        ?.getLastMessageContent()
                        ?: timelineEvent.getLastMessageContent()
```

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/render/ProcessBodyOfReplyToEventUseCase.kt [hunk 3]
```diff
@@ -100,10 +103,23 @@ class ProcessBodyOfReplyToEventUseCase @Inject constructor(
                         repliedToEvent.isPollEnd() -> stringProvider.getString(R.string.message_reply_to_sender_ended_poll)
                         else -> ""
                     }
+                    val repliedText = when {
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/render/ProcessBodyOfReplyToEventUseCase.kt:106]: Maybe to have something easier to read, we should avoid nesting the when and handle `repliedToEvent.isPollStart()` and `repliedToEvent.isPollEnd()` separately.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt [hunk 4]
```diff
@@ -754,6 +766,7 @@ internal class LocalEchoEventFactory @Inject constructor(
             MessageType.MSGTYPE_POLL_START -> {
                 return TextContent((content as? MessagePollContent)?.getBestPollCreationInfo()?.question?.getBestQuestion() ?: "")
             }
+            MessageType.MSGTYPE_POLL_END -> return TextContent("Ended poll")
```
Comment (mnaturel) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt:769]: Maybe we should use lower case for the start like `ended poll` to follow convention of other contents?

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/render/ProcessBodyOfReplyToEventUseCase.kt [hunk 5]
```diff
@@ -100,10 +103,23 @@ class ProcessBodyOfReplyToEventUseCase @Inject constructor(
...
+                        repliedToEvent.isPollEnd() -> {
```
Comment (mnaturel) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/render/ProcessBodyOfReplyToEventUseCase.kt:107]: I think we should add a new unit test to cover the end poll event case.

**Review by mnaturel — COMMENTED** [review]
I have added some remarks and questions. Also, when testing it, I have seen 2 things: 
- I wonder how should behave the preview of the message when replying to end message. Should we also have the poll title for consistency?

- When editing a reply it no more displays the poll title but the fallback. This is linked to one of my remarks.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt (hunk 1 above)
Comment (onurays) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt:615]: Make sense, but need testing.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt (hunk 2 above)
Comment (onurays) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt:730]: I agree that last message may also be null.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt (hunk 4 above)
Comment (onurays) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt:769]: Design for this suggests to use uppercase but I wouldn't mind to make it lowercase.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt (hunk 1 above)
Comment (mnaturel) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt]: Fixed in [7a95ece](https://github.com/vector-im/element-android/pull/7969/commits/7a95ece7381c17802e718f6473ca574fea5d621a)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt (hunk 2 above)
Comment (mnaturel) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/send/LocalEchoEventFactory.kt:731]: Added fallback in [8fb38f5](https://github.com/vector-im/element-android/pull/7969/commits/8fb38f57dcab7f30f9c781bde42165bbb0dff797)

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/render/ProcessBodyOfReplyToEventUseCase.kt (hunk 3 above)
Comment (mnaturel) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/render/ProcessBodyOfReplyToEventUseCase.kt]: Refactored in [5820185](https://github.com/vector-im/element-android/pull/7969/commits/58201851ad196978f00df96bf1b76257a242c256)

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/render/ProcessBodyOfReplyToEventUseCase.kt (hunk 5 above)
Comment (mnaturel) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/render/ProcessBodyOfReplyToEventUseCase.kt]: Unit tests updated in [7683b83](https://github.com/vector-im/element-android/pull/7969/commits/7683b8325c3535af3b478d3759ac7e00431c05fd)

**Review by mnaturel — APPROVED** [review]
LGTM.

**Pull request merged**